lib/util/cache.js

```diff
@@ -0,0 +1,63 @@
+/**
+ * @fileoverview Cache object that manages multiple namespaced collections of data.
+ * @author Kai Cataldo
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Public Interface
+//------------------------------------------------------------------------------
+
+/**
+ * Cache
+ *
+ * A cache object that manages multiple namespaced collections of data.
+ * Each namespaced collection is a WeakMap instance.
+ * The cache will lazily create a new cache for the specified collection namespace the first
+ * time set() is called with a new collection type.
+ */
+module.exports = class Cache {
```

Comment:
Have you run performance tests on this? Since it stores every result from `getTokensBefore` etc. from a file, this intuitively seems like it would use a very large amount of memory. Maybe we should impose a maximum capacity and drop things from the cache if they haven't been used for awhile.

Also, this could cause confusing results if a rule accidentally mutates a result (e.g. `sourceCode.getTokensBefore(foo).reverse().forEach()`, because if another rule calls `getTokensBefore` afterwards on the same list, it won't be correct.